src/rules/rule-builder.ts

```diff
@@ -0,0 +1,210 @@
+import {Example, Options, Rule, RuleType, registerRule, LinterSettings} from '../rules';
+import {BooleanOption, DropdownOption, DropdownRecord, MomentFormatOption, Option, TextAreaOption, TextOption} from '../option';
+import {logDebug} from '../logger';
+
+export abstract class RuleBuilderBase {
+  static #ruleMap = new Map<string, Rule>();
+
+  static getRule<TOptions extends Options>(this: (new() => RuleBuilder<TOptions>)): Rule {
+    if (!RuleBuilderBase.#ruleMap.has(this.name)) {
+      const builder = new this();
+      const rule = new Rule(builder.name, builder.description, builder.type, builder.safeApply.bind(builder), builder.exampleBuilders.map((b) => b.example), builder.optionBuilders.map((b) => b.option), builder.isSpecial);
+      RuleBuilderBase.#ruleMap.set(this.name, rule);
+    }
+
+    return RuleBuilderBase.#ruleMap.get(this.name);
+  }
+
+  static applyIfEnabledBase(rule: Rule, text: string, settings: LinterSettings, extraOptions: Options): [result: string, isEnabled: boolean] {
+    const optionsFromSettings = rule.getOptions(settings);
+    if (optionsFromSettings[rule.enabledOptionName()]) {
+      const options = Object.assign({}, optionsFromSettings, extraOptions) as Options;
+      logDebug(`Running ${rule.name}`);
+      return [rule.apply(text, options), true];
+    } else {
+      return [text, false];
+    }
+  }
+}
+
+export default abstract class RuleBuilder<TOptions extends Options> extends RuleBuilderBase {
+  abstract get OptionsClass(): (new() => TOptions);
+
+  static register<TOptions extends Options>(RuleBuilderClass: typeof RuleBuilderBase & (new() => RuleBuilder<TOptions>)): void {
+    const rule = RuleBuilderClass.getRule();
+    registerRule(rule);
+  }
+
+  safeApply(text: string, options?: Options): string {
+    return this.apply(text, this.buildRuleOptions(options));
+  }
+
+  buildRuleOptions(options?: Options): TOptions {
+    options = options ?? {};
+    const defaultOptions = new this.OptionsClass();
+    const ruleOptions = Object.assign(defaultOptions, options) as TOptions;
+
+    for (const optionBuilder of this.optionBuilders) {
+      optionBuilder.setRuleOption(ruleOptions, options);
+    }
+
+    return ruleOptions;
+  }
+
+  abstract get name(): string;
+  abstract get description(): string;
+  abstract get type(): RuleType;
+  abstract apply(text: string, options: TOptions): string;
+  abstract get exampleBuilders(): ExampleBuilder<TOptions>[];
+  abstract get optionBuilders(): OptionBuilderBase<TOptions>[];
+  get isSpecial(): boolean {
```

Comment:
I might rename this to `isRunBeforeOrAfterOtherRules` as it makes it clear what is special about this method.